### PR TITLE
RavenDB-6780: Allocations and small profile based optimizations

### PR DESCRIPTION
**Pull request by redknightlois**

### CODE REVIEW

**Comment by ayende** [comment]
This gives us a lot of failing tests.

File: src/Sparrow/Json/JsonOperationContext.cs [hunk 1]
```diff
@@ -273,7 +269,9 @@ public virtual void Dispose()
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public LazyStringValue GetLazyStringForFieldWithCaching(StringSegment key)
-        {
+        {            
+            Debug.Assert(key.Length == key.String.Length, "If the segment do not comprise the whole string this method will fail.");
```
Comment (ayende) [inline src/Sparrow/Json/JsonOperationContext.cs:273]: This is raised by a bunch of tests (run FastTests).
This also indicate that we need to consider whatever we are actually saving anything by using StringSegment or if we should remove it.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Sparrow/Json/JsonOperationContext.cs (hunk 1 above)
Comment (redknightlois) [reply src/Sparrow/Json/JsonOperationContext.cs:273]: Mhhh that makes sense I run tests on release mode. I added that as a safeguard, probably it is more complex, you can remove it without problem

Comment (ayende) [reply src/Sparrow/Json/JsonOperationContext.cs:273]: Not really, no, this is actually indicative of an issue.

**Pull request closed without merging**